File: Leetcode/Medium/105_construct_binary_tree.py
# Definition for a binary tree node.
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return None

        root = TreeNode(val=preorder[0])
        mid = inorder.index(preorder[0])
        root.left = self.buildTree(
            preorder=preorder[1 : mid + 1], inorder=inorder[:mid]
        )
        root.right = self.buildTree(
            preorder=preorder[mid + 1 :], inorder=inorder[mid + 1 :]
        )

        return root


def traversalTree(root: Optional[TreeNode]):
    if not root:
        return

    print(root.val)
    traversalTree(root.left)
    traversalTree(root.right)


if __name__ == "__main__":
    preorder: list[int] = [3, 9, 20, 15, 7]
    inorder: list[int] = [9, 3, 15, 20, 7]

    solution = Solution()
    result = solution.buildTree(preorder, inorder)
    traversalTree(result)

File: Leetcode/Medium/105_construct_binary_tree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func BuildTree(preoder, inorder []int) *TreeNode {
	if len(preoder) <= 0 || len(inorder) <= 0 {
		return nil
	}

	root := &TreeNode{Val: preoder[0]}
	mid := 0
	for idx, val := range inorder {
		if val == preoder[0] {
			mid = idx
			break
		}
	}

	root.Left = BuildTree(preoder[1:mid+1], inorder[:mid])
	root.Right = BuildTree(preoder[mid+1:], inorder[mid+1:])

	return root

}

func Preorder(root *TreeNode) {
	if root == nil {
		return
	}
	fmt.Println(root.Val)
	Preorder(root.Left)
	Preorder(root.Right)
}

func main() {
	// preorder := []int{3, 9, 20, 15, 7}
	// inorder := []int{9, 3, 15, 20, 7}

	// result := BuildTree(preorder, inorder)
	// Preorder(result)

	// Note : 切Slice的一些注意事項
	// a := []int{1, 2, 3} // len(a)==3 cap(a)==3
	// fmt.Println(a[3:]) // 如果起始位置為len + 1 則result -> []
	// fmt.Println(a[3:4]) // slice bounds out of range [:4] with capacity 3
}
